### DISCUSSION TITLE
How to implement software render windows?

### DISCUSSION THREAD

**Discussion by sudongg** [AvaloniaUI/Avalonia#6328] [Q&A]
hi,
On WPF, I can set the rendering mode of the window to software through the following code: 
```
        protected override void OnSourceInitialized(EventArgs e)
        {
            var hwndSource = PresentationSource.FromVisual(this) as HwndSource;

            if (hwndSource != null)
            {
                hwndSource.CompositionTarget.RenderMode = RenderMode.SoftwareOnly;
            }

            base.OnSourceInitialized(e);
        }
```
How to implement this function in Avalonia?

**Comment by kekekeks** [comment]
Why exactly would you want to have a software-rendered window? While it's technically possible to do with the current version, it's not a supported scenario (software rendering can be enable application-wide though). We'd like to know scenarios when it's needed to consider adding per-window support for software rendering.


**Note:** this is a **huge hack** and might stop working in future versions of Avalonia, use on your own risk

1) Create something like:
```cs
class SoftwareWindowImplProxy : IWindowImpl
{
   IWindowImpl _real;
   ...
}
```
that forwards everything to `_real`

2) in `IEnumerable<object> Surfaces { get; }` filter out everything but `IFramebufferPlatformSurface` (it should be the last one)

3) pass your proxy to `Window`'s ctor

i. e.
```cs
class MySoftwareRenderedWindow
{
     public MySoftwareRenderedWindow() : base(new SoftwareWindowImplProxy(PlatformManager.CreateWindow()))
}
```

**Reply by sudongg** [reply]
Hi，kekekeks,
I am porting a wpf player project to Avalonia.
This player has two layers of windows, the bottom window is used to play the video (NativeControlHost), and the upper layer uses a transparent window to draw the UI.

In most cases, the upper window can draw ui very well (both the display quality and the number of frames are better than WPF).
However, if the GUP usage exceeds 65 percent when playing a video, The number of UI frames decreased significantly.
Wpf also has the same situation. The solution is that the upper window only uses software for rendering. In this way, even if Gup is used more than 65%, it will not affect the number of frames.

According to the answer you gave, Avalonia currently has no easy way to enable software rendering.  I haven't found any related Interface in IWindowImpl, and I don't know how IFramebufferPlatformSurface works.
I hope Avalonia can provide a solution similar to wpf for software rendering of windows.

**Comment by maxkatz6** [comment]
You can disable software rendering on windows for whole avalonia application (not for a single specific window) with AllowEglInitialization. In your Program.cs it should be something like this:
```
return AppBuilderBase<TAppBuilder>.Configure<App>()
                .UsePlatformDetect()
                .With(new Win32PlatformOptions
                {
                    AllowEglInitialization = false
                })
```

Probably it's enough for you, if you don't need crossplatform (unless other platforms have the same problem) and if your "bottom" window is just a video player.

**Reply by sudongg** [reply]
Thank you, AllowEglInitialization = false is effective.

I did a test,
Add 80 textblocks in the upper window and repeat the translatetransform animation.

In the case of AllowEglInitialization = true,
When not playing:   the fps is 59-60, and the animation is smooth.
When playing:          the fps is 47-54, GPU use 87%, CPU use 9%. you can feel obvious freeze.

In the case of AllowEglInitialization = false,
When not playing:   the fps is 60-63.
However, although the frame rate is relatively high, for me, the visual effect when AllowEglInitialization = true is not as smooth as Hardware rendering.
When playing:          the fps is 60-62, GPU use 74%,  CPU use 8%.


When using software rendering, GPU usage has almost no effect on the number of frames. But I can feel a slight jitter in the animation.
I tried to reduce the number of TextBlocks in Avalonia and used a custom RenderTimer, but nothing improved.
If this problem cannot be solved, my project may not continue.

If it can be stabilized at 60 frames when using software rendering on wpf, there is no difference between animation fluency and hardware rendering.
It's just that when using software rendering on wpf it is easy to fall below 60 frames (such as 20 TranslateTransform animations at the same time), and using software rendering on Avalonia can maintain 60fps even if 80 TextBlocks are animated at the same time.

**Reply by Gillibald** [reply]
Text rendering is always done by CPU. Even with software rendering you have to transfer something to GPU at some point.

**Reply by sudongg** [reply]
> Text rendering is always done by CPU. Even with software rendering you have to transfer something to GPU at some point.

In WPF, you can convert text into Geometry, then use RenderTargetBitmap to render Geometry into a Bitmap, and use DrawingContext for rendering.
I have used it in a project, and the animation works well after converting it into a Bitmap. And the number of frames is higher than using TextBlocks directly.
Is there a similar implementation on Avalonia?
Maybe it can solve the problem that the software rendering is not smooth when using TextBlocks for animation.

**Reply by maxkatz6** [reply]
I believe this PR allows geometry creating from text https://github.com/AvaloniaUI/Avalonia/pull/4893
Help with testing might be useful for @Gillibald

**Reply by maxkatz6** [reply]
Ok, it seems it wasn't yet implemented. Issue https://github.com/AvaloniaUI/Avalonia/issues/6043

**Reply by Gillibald** [reply]
I need more information about how you animate text to give you some advice

**Reply by Gillibald** [reply]
WPF's RenderTargetBitmap is hardware accelerated. Avalonia's version is currently software only and requires transferring data from CPU to GPU bound memory.

**Reply by sudongg** [reply]
> WPF's RenderTargetBitmap is hardware accelerated. Avalonia's version is currently software only and requires transferring data from CPU to GPU bound memory.

The animation moving from right to left is like this:
https://github.com/sudongg/UnrealTournament/blob/clean-master/1.png
I have achieved smooth animation on WPF even when the GPU is used more than 75%.

I only used a doubleAnimate to keep the control moving, and draw all the text on the control, and only need to update the rendering when the text is added or exceeded. And the software window must be set, otherwise the animation will not be smooth. 

However, setting the software window plus a transparent window requires strict control of the number of text on WPF to maintain a smooth animation.

On Avalonia, I can only use TextBlocks and set animations for each one. When the GPU usage is less than 60%, everything is fine. I am satisfied, but once it exceeds, the animation is not smooth. 
Even if AllowEglInitialization = false is set, there is no improvement.

**Reply by Gillibald** [reply]
You should be able to use a DrawingImage to achieve what you need. https://github.com/AvaloniaUI/Avalonia/blob/master/samples/RenderDemo/Pages/GlyphRunPage.xaml.cs So in the end you will be animating an Image control.

**Reply by Gillibald** [reply]
The GlyphRun itself only has to change when the text changes

**Reply by Gillibald** [reply]
To get a shaped GlyphRun you can use the TextShaper

**Reply by sudongg** [reply]
> To get a shaped GlyphRun you can use the TextShaper

For this demo, there are several questions
1. How to set the position of GlyphRun?
2. How to modify the font family?
3. How to get GlyphRun outline?

**Reply by Gillibald** [reply]
1. Move the DrawingImage
2. Construct a new typeface and reshape - > new GlyphRun
3. Why do you need the outline?

**Reply by sudongg** [reply]
> 1. Move the DrawingImage
> 2. Construct a new typeface and reshape - > new GlyphRun
> 3. Why do you need the outline?

Please refer to the picture I gave:
https://github.com/sudongg/UnrealTournament/blob/clean-master/1.png
We call it barrage or Bullet screen, which is used to display user comments in real time while the video is playing.
Quite popular in our country.

I need to draw multiple texts, maybe more than 40. If I just set the position of the DrawingImage, then I need to put 40 DrawingImages in the control.

Please explain in more detail how to set the font family. I have experience in drawing GlyphRun and outline on D2D, but I can't find a similar interface on Avalonia.

Please refer to the picture above as to why the outline is drawn. If the outline is not drawn, the white font will be difficult to recognize on the light-colored screen.

**Reply by maxkatz6** [reply]
@sudongg this picture is permissions limited

**Reply by sudongg** [reply]
> @sudongg this picture is permissions limited

Please try this link： https://www.dropbox.com/s/uwhbm3s2lhiv83a/1.png?dl=0

**Reply by Gillibald** [reply]
So you just display text. `TextLayout` should work for you. That can be used directly via DrawingContext.

**Comment by Gillibald** [comment]
```
using (var contextImpl = renderTargetBitmap.CreateDrawingContext(null))
using (var context = new DrawingContext(contextImpl, false))
{
    var typeface = Typeface.Default; //new Typeface("MyFont")

    //If nothing changes related to the TextLayout this can be cached
    var textLayout = new TextLayout("My text", typeface, fontSize: 12, foreground: Brushes.White);

    //Change the origin to draw the layout at different positions.
    textLayout.Draw(context, new Point(0, 0));//This signatur is only available in current master. Use a transform if you can't use a nightly build.
}
```

**Reply by Gillibald** [reply]
RenderTargetBitmap is software mode only so this should be what you want.

**Reply by sudongg** [reply]
> RenderTargetBitmap is software mode only so this should be what you want.

thank you,
I convert each textlayout into rendertargetbitmap and cache it, and then just draw the bitmap and set the point.
However, there is no outline text yet, so the test cannot be completed. Do I have to wait for the outline text to complete?

**Reply by Gillibald** [reply]
If you can't wait it should be possible to use https://docs.microsoft.com/en-us/dotnet/api/skiasharp.skfont.getglyphpaths?view=skiasharp-2.80.2#SkiaSharp_SKFont_GetGlyphPaths_System_ReadOnlySpan_System_UInt16__SkiaSharp_SKGlyphPathDelegate_ to convert the TextLayout's TextRuns to path elements.

**Reply by Gillibald** [reply]
Each TextLine holds Shaped(TextRuns) and each ShapedTextRun holds a GlyphRun. A GlyphRuns glyph indices can be used to optain SKPath elements.

**Reply by Gillibald** [reply]
For convenience there is also this method on SKPaint https://docs.microsoft.com/en-us/dotnet/api/skiasharp.skpaint.gettextpath?view=skiasharp-2.80.2#SkiaSharp_SKPaint_GetTextPath_System_IntPtr_System_IntPtr_System_Single_System_Single_ just make sure to set the TextEncoding

**Reply by Gillibald** [reply]
My current plan is to port https://docs.microsoft.com/en-us/dotnet/api/system.windows.media.glyphrun.buildgeometry?view=net-5.0 and https://docs.microsoft.com/en-us/dotnet/api/system.windows.media.glyphtypeface.getglyphoutline?view=net-5.0

That will be the main building block to implement https://docs.microsoft.com/en-us/dotnet/api/system.windows.media.formattedtext.buildgeometry?view=net-5.0 and https://docs.microsoft.com/en-us/dotnet/api/system.windows.media.formattedtext.buildhighlightgeometry?view=net-5.0

FormattedText is already ported here: https://github.com/AvaloniaUI/Avalonia/pull/4893
